File: backend/permissions.py
"""
RBAC — Role-Based Access Control — LabControl UTECAN
=====================================================
Fuente única de verdad para permisos del sistema.

Cada permiso tiene la forma  "recurso:acción"  y mapea a los roles
que pueden ejecutarlo. Todos los routers deben usar require_permission()
en vez de require_roles() con roles hardcodeados.

Roles disponibles:
  SUPER_ADMIN  → control total institucional
  LAB_ADMIN    → control de su laboratorio asignado
  ADMINISTRATIVO → gestión de comunicados de su departamento
  DOCENTE      → reservaciones y sesiones de clase propias

Convención de acciones:
  :read    → GET lista o detalle
  :write   → POST / PUT / PATCH crear o editar
  :delete  → DELETE / dar de baja
  :admin   → acciones privilegiadas (resolver conflictos, bloquear, etc.)
  :own     → cualquier rol autenticado accede solo a sus propios recursos
  :import  → importación masiva desde Excel
"""

from fastapi import Depends, HTTPException, status
from models.usuario import RolUsuario, Usuario
from dependencies import get_current_user

# Alias cortos para leer la matriz más fácil
SA  = RolUsuario.SUPER_ADMIN
LA  = RolUsuario.LAB_ADMIN
AD  = RolUsuario.ADMINISTRATIVO
DO  = RolUsuario.DOCENTE

# ── Matriz de permisos ─────────────────────────────────────────────────────────
#
# PERMISSIONS[permiso] = frozenset de roles que lo tienen
#
PERMISSIONS: dict[str, frozenset[RolUsuario]] = {

    # ── Laboratorios ────────────────────────────────────────────────────────────
    "laboratorios:read":        frozenset({SA, LA, DO}),   # ver lista y detalle
    "laboratorios:write":       frozenset({SA}),            # crear / editar lab
    "laboratorios:delete":      frozenset({SA}),            # desactivar lab
    "pcs:read":                 frozenset({SA, LA, DO}),   # ver PCs
    "pcs:write":                frozenset({SA, LA}),        # agregar / editar PCs
    "pcs:admin":                frozenset({SA, LA}),        # marcar BAJA / estado

    # ── Usuarios ────────────────────────────────────────────────────────────────
    "usuarios:read":            frozenset({SA, LA}),        # listar usuarios
    "usuarios:write":           frozenset({SA}),            # crear / editar usuario
    "usuarios:delete":          frozenset({SA}),            # desactivar
    "usuarios:reset":           frozenset({SA}),            # reset de contraseña
    "usuarios:self":            frozenset({SA, LA, AD, DO}),   # cambiar propia contraseña
    "usuarios:import":          frozenset({SA}),            # importar desde Excel

    # ── Horarios ────────────────────────────────────────────────────────────────
    "horarios:read":            frozenset({SA, LA, DO}),   # ver slots disponibles
    "horarios:write":           frozenset({SA, LA}),        # crear / editar slots
    "horarios:delete":          frozenset({SA, LA}),        # desactivar slot

    # ── Reservaciones ───────────────────────────────────────────────────────────
    "reservaciones:read":       frozenset({SA, LA, DO}),   # ver reservaciones
    "reservaciones:write":      frozenset({SA, LA, DO}),   # crear / editar propia
    "reservaciones:admin":      frozenset({SA, LA}),        # resolver conflictos, bloquear

    # ── Sesiones de clase ───────────────────────────────────────────────────────
    "sesiones:read":            frozenset({SA, LA, DO}),   # ver sesiones
    "sesiones:write":           frozenset({SA, LA, DO}),   # abrir / cerrar sesión propia
    "sesiones:admin":           frozenset({SA, LA}),        # ver y cerrar cualquier sesión
    "sesiones:asignar":         frozenset({SA, LA, DO}),   # asignar alumno a PC
    "sesiones:incidencia":      frozenset({SA, LA, DO}),   # reportar incidencia en sesión

    # ── Inventario ──────────────────────────────────────────────────────────────
    "inventario:read":          frozenset({SA, LA, DO}),   # ver activos
    "inventario:write":         frozenset({SA, LA}),        # crear / editar activo
    "inventario:delete":        frozenset({SA, LA}),        # dar de baja activo
    "inventario:import":        frozenset({SA, LA}),        # importar Excel

    # ── Préstamos ────────────────────────────────────────────────────────────────
    "prestamos:read":           frozenset({SA, LA, DO}),   # ver préstamos
    "prestamos:write":          frozenset({SA, LA, DO}),   # registrar préstamo
    "prestamos:devolver":       frozenset({SA, LA}),        # registrar devolución

    # ── Mantenimiento ────────────────────────────────────────────────────────────
    "mantenimiento:read":       frozenset({SA, LA, DO}),   # ver incidentes / preventivos
    "mantenimiento:write":      frozenset({SA, LA}),        # crear / editar
    "mantenimiento:delete":     frozenset({SA, LA}),        # eliminar preventivo

    # ── Incidentes ───────────────────────────────────────────────────────────────
    "incidentes:read":          frozenset({SA, LA, DO}),
    "incidentes:write":         frozenset({SA, LA, DO}),   # cualquiera puede reportar
    "incidentes:admin":         frozenset({SA, LA}),        # cambiar estado / resolver

    # ── Catálogo académico ───────────────────────────────────────────────────────
    "catalogo:read":            frozenset({SA, LA, DO}),   # buscar alumnos / materias
    "catalogo:write":           frozenset({SA, LA}),        # crear / editar
    "catalogo:delete":          frozenset({SA, LA}),        # desactivar
    "catalogo:import":          frozenset({SA, LA}),        # importar Excel

    # ── Reportes ─────────────────────────────────────────────────────────────────
    "reportes:read":            frozenset({SA, LA}),        # ver reportes mensuales
    "reportes:export":          frozenset({SA, LA}),        # descargar Excel

    # ── Notificaciones ───────────────────────────────────────────────────────────
    "notificaciones:own":       frozenset({SA, LA, AD, DO}),   # ver / marcar las propias

    # Departamentos y comunicados institucionales
    "departamentos:read":       frozenset({SA, LA, AD, DO}),
    "departamentos:write":      frozenset({SA}),
    "comunicados:own":          frozenset({SA, LA, AD, DO}),
    "comunicados:write":        frozenset({SA, LA, AD}),
}


# ── Helper: consulta puntual desde código ─────────────────────────────────────

def can(rol: RolUsuario, permiso: str) -> bool:
    """
    Verifica en tiempo de ejecución si un rol tiene un permiso.

    Ejemplo:
        if can(usuario.rol, "reportes:read"):
            ...
    """
    return rol in PERMISSIONS.get(permiso, frozenset())


# ── Dependency factory para FastAPI ───────────────────────────────────────────

def require_permission(permiso: str):
    """
    Dependency que lanza HTTP 403 si el usuario no tiene el permiso requerido.

    Uso en routers:
        @router.post("/activos")
        def crear_activo(
            data: ActivoCreate,
            current_user: Usuario = Depends(require_permission("inventario:write"))
        ):
            ...

    Ventaja sobre require_roles(): el permiso queda documentado en la firma
    del endpoint y en la matriz PERMISSIONS — no hay que buscar en el código
    qué roles están permitidos.
    """
    allowed = PERMISSIONS.get(permiso, frozenset())

    if not allowed:
        # Permiso no definido en la matriz → error de programación, no de runtime
        raise ValueError(
            f"Permiso '{permiso}' no existe en la matriz PERMISSIONS. "
            f"Agrégalo en backend/permissions.py"
        )

    def _check(current_user: Usuario = Depends(get_current_user)) -> Usuario:
        if current_user.rol not in allowed:
            roles_str = ", ".join(r.value for r in allowed)
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail={
                    "error":     "Acceso denegado",
                    "permiso":   permiso,
                    "tu_rol":    current_user.rol.value,
                    "roles_permitidos": [r.value for r in allowed],
                }
            )
        return current_user

    return _check


# ── Endpoint de introspección (solo SUPER_ADMIN) ───────────────────────────────

def get_permission_matrix() -> dict:
    """
    Devuelve la matriz completa de permisos serializada.
    Útil para el endpoint GET /rbac/permissions.
    """
    return {
        permiso: sorted(r.value for r in roles)
        for permiso, roles in sorted(PERMISSIONS.items())
    }

File: frontend/src/config/permissions.js
/**
 * RBAC — Matriz de permisos del frontend
 * SIGA UTECAN
 *
 * Espejo del archivo backend/permissions.py
 * Mantener sincronizados cuando se agreguen módulos.
 *
 * Uso:
 *   import { can, ROUTE_PERMISSIONS } from '../config/permissions';
 *   if (can(usuario.rol, 'reportes:read')) { ... }
 */

// ── Roles ──────────────────────────────────────────────────────────────────────
export const ROLES = {
  SUPER_ADMIN: 'SUPER_ADMIN',
  LAB_ADMIN:   'LAB_ADMIN',
  RESPONSABLE_LAB: 'RESPONSABLE_LAB',
  ADMINISTRATIVO: 'ADMINISTRATIVO',
  TUTORIA_ADMIN: 'TUTORIA_ADMIN',
  SERVICIOS_ESCOLARES: 'SERVICIOS_ESCOLARES',
  ALUMNO: 'ALUMNO',
  DOCENTE:     'DOCENTE',
};

const SA = ROLES.SUPER_ADMIN;
const LA = ROLES.LAB_ADMIN;
const RL = ROLES.RESPONSABLE_LAB;
const AD = ROLES.ADMINISTRATIVO;
const TA = ROLES.TUTORIA_ADMIN;
const SE = ROLES.SERVICIOS_ESCOLARES;
const AL = ROLES.ALUMNO;
const DO = ROLES.DOCENTE;

// ── Matriz de permisos ─────────────────────────────────────────────────────────
// permiso → array de roles que lo tienen
export const PERMISSIONS = {
  // Laboratorios
  'laboratorios:read':      [SA, LA, RL, DO],
  'laboratorios:write':     [SA],
  'laboratorios:delete':    [SA],
  'pcs:read':               [SA, LA, DO],
  'pcs:write':              [SA, LA],
  'pcs:admin':              [SA, LA],

  // Usuarios
  'usuarios:read':          [SA, LA],
  'usuarios:write':         [SA],
  'usuarios:delete':        [SA],
  'usuarios:reset':         [SA],
  'usuarios:self':          [SA, LA, RL, AD, TA, SE, AL, DO],
  'usuarios:import':        [SA],

  // Horarios
  'horarios:read':          [SA, LA, DO],
  'horarios:write':         [SA, LA],
  'horarios:delete':        [SA, LA],

  // Reservaciones
  'reservaciones:read':     [SA, LA, DO],
  'reservaciones:write':    [SA, LA, DO],
  'reservaciones:admin':    [SA, LA],

  // Sesiones
  'sesiones:read':          [SA, LA, DO],
  'sesiones:write':         [SA, LA, DO],
  'sesiones:admin':         [SA, LA],
  'sesiones:asignar':       [SA, LA, DO],
  'sesiones:incidencia':    [SA, LA, DO],

  // Inventario
  'inventario:read':        [SA, LA, RL, AD, DO],
  'inventario:write':       [SA, LA, RL],
  'inventario:delete':      [SA, LA, RL],
  'inventario:import':      [SA, LA, RL],

  // Préstamos
  'prestamos:read':         [SA, LA, RL, DO],
  'prestamos:write':        [SA, LA, RL, DO],
  'prestamos:devolver':     [SA, LA, RL],

  // Mantenimiento
  'mantenimiento:read':     [SA, LA, DO],
  'mantenimiento:write':    [SA, LA],
  'mantenimiento:delete':   [SA, LA],

  // Incidentes
  'incidentes:read':        [SA, LA, DO],
  'incidentes:write':       [SA, LA, DO],
  'incidentes:admin':       [SA, LA],

  // Catálogo
  'catalogo:read':          [SA, LA, SE, DO],
  'catalogo:write':         [SA, SE],
  'catalogo:delete':        [SA, SE],
  'catalogo:import':        [SA, SE],

  // Reportes
  'reportes:read':          [SA, LA],
  'reportes:export':        [SA, LA],
  'sistema:backup':         [SA],

  // Notificaciones
  'notificaciones:own':     [SA, LA, RL, AD, TA, SE, AL, DO],
  'comunicados:own':        [SA, LA, RL, AD, TA, SE, AL, DO],
  'comunicados:write':      [SA, LA, TA],
  'departamentos:read':     [SA, LA, RL, AD, TA, SE, DO],
  'departamentos:write':    [SA],
  'tutoria:admin':          [SA, TA],
  'tutoria:own':            [DO],
};

// ── Permisos requeridos por ruta ───────────────────────────────────────────────
// Fuente de verdad para RutaProtegida en App.jsx
export const ROUTE_PERMISSIONS = {
  '/admin':                       [SA],
  '/lab':                         [SA, LA],
  '/administrativo':              [AD],
  '/servicios-escolares':         [SA, SE],
  '/servicios-escolares/alumnos': [SA, SE],
  '/servicios-escolares/estudios-socioeconomicos': [SA, SE],
  '/alumno':                      [AL],
  '/alumno/estudio-socioeconomico': [AL],
  '/admin/tutoria':               [SA, TA],
  '/admin/laboratorios':          [SA, LA, RL],
  '/admin/laboratorios/:labId':   [SA, LA, RL],
  '/admin/usuarios':              [SA],
  '/admin/departamentos':         [SA, AD],
  '/admin/horarios':              [SA, LA],
  '/admin/reservaciones':         [SA, LA, DO],
  '/admin/inventario':            [SA, LA, RL, AD],
  '/admin/prestamos':             [SA, LA, RL],
  '/admin/mantenimiento':         [SA, LA],
  '/admin/catalogo':              [SA],
  '/admin/reportes':                      [SA, LA],
  '/admin/historial-alumno':              [SA],
  '/admin/respaldos':                      [SA],
  '/admin/sesion/:sesionId':              [SA, LA, DO],
  '/admin/sesion/:sesionId/asistencia':   [SA, LA, DO],
  '/docente':                             [DO],
  '/docente/reservaciones':       [DO],
  '/docente/sesion/:sesionId':    [SA, LA, DO],
  '/comunicados':                 [SA, LA, AD, TA, DO],
  '/admin/comunicados':           [SA, LA, AD, TA],
  '/medico/consultorio':           ['MEDICO', SA],
};

// ── Helpers ────────────────────────────────────────────────────────────────────

/**
 * Verifica si un rol tiene un permiso.
 *
 * @param {string} rol   - 'SUPER_ADMIN' | 'LAB_ADMIN' | 'DOCENTE'
 * @param {string} perm  - clave de PERMISSIONS, ej. 'inventario:write'
 * @returns {boolean}
 *
 * @example
 *   import { can } from '../config/permissions';
 *   const puedeEditar = can(usuario.rol, 'inventario:write');
 */
export function can(rol, perm) {
  const allowed = PERMISSIONS[perm];
  if (!allowed) {
    console.warn(`[RBAC] Permiso desconocido: "${perm}"`);
    return false;
  }
  return allowed.includes(rol);
}

/**
 * Devuelve todos los permisos de un rol.
 *
 * @param {string} rol
 * @returns {string[]}
 */
export function getPermissions(rol) {
  return Object.entries(PERMISSIONS)
    .filter(([, roles]) => roles.includes(rol))
    .map(([perm]) => perm);
}

/**
 * Verifica si un rol tiene acceso a una ruta dada.
 *
 * @param {string} rol
 * @param {string} path  - ruta de React Router, ej. '/admin/reportes'
 * @returns {boolean}
 */
export function canAccessRoute(rol, path) {
  const allowed = ROUTE_PERMISSIONS[path];
  if (!allowed) return true; // ruta pública o no controlada
  return allowed.includes(rol);
}
